### PR TITLE
Refactor field validation

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR mainly refactors the code for fields validation.

The changes:
- Previously, value to validate was passed as a raw `Object`. Now, the value is wrapped in `FieldValue`.
- `io.spine.code.proto.FieldDeclaration` was renamed to `FieldDeclarationProto`.
- `FieldDeclaration` was moved from `io.spine.validate` to `io.spine.code.proto`.
- `MessageValue` was introduced to wrap a `Message` to validate, in the same manner as `FieldValue`.

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@dmdashenkov PTAL.

File: base/src/main/java/io/spine/code/proto/FieldDeclaration.java [hunk 1]
```diff
@@ -20,39 +20,117 @@
...
+        return entityKind.getNumber() > 0;
+    }
+
+    private boolean matchesIdName() {
+        String name = field.getName();
+        return "id".equals(name) || name.endsWith("_id");
+    }
+
+    private boolean isFirstField() {
+        return field.getIndex() == 0;
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/code/proto/FieldDeclaration.java]: Shouldn't the field index be `1`?

File: base/src/main/java/io/spine/code/proto/FieldDeclarationProto.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                                 AbstractMessageDeclaration originMessage) {
+        this.descriptor = descriptor;
+        this.originMessage = originMessage;
+    }
+
+    public FieldName name() {
+        return FieldName.of(descriptor);
+    }
+
+    public FieldDescriptorProto descriptor() {
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/code/proto/FieldDeclarationProto.java]: Please document public members.

File: base/src/main/java/io/spine/code/proto/FieldDeclarationProto.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+package io.spine.code.proto;
+
+import com.google.protobuf.DescriptorProtos.FieldDescriptorProto;
+
+import java.util.Optional;
+
+/**
+ * A declaration of a message {@linkplain FieldDescriptorProto field}.
+ */
+public class FieldDeclarationProto {
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/code/proto/FieldDeclarationProto.java]: Can this be final?

File: base/src/main/java/io/spine/validate/FieldDeclaration.java [hunk 4]
```diff
@@ -98,13 +76,41 @@ boolean isCommandId() {
...
-    boolean isNotRepeatedOrMap() {
-        return !isRepeated()
-                && !field.isMapField();
+    /**
+     * Determines whether the declaration is a scalar type.
+     *
+     * @return {@code true} if the declaration neither map nor repeated, {@code false} otherwise
+     */
+    boolean isScalar() {
+        return !isMap() && !isRepeated();
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/validate/FieldDeclaration.java]: This is a duplicate.

File: base/src/main/java/io/spine/validate/FieldDeclaration.java [hunk 5]
```diff
@@ -20,54 +20,32 @@
...
  *
  * <p>The field can be declared in a message or enum.
  *
- * <p>Unlike {@link io.spine.code.proto.FieldDeclaration}, the class uses
+ * <p>Unlike {@link io.spine.code.proto.FieldDeclarationProto}, the class uses
  * {@link com.google.protobuf.Descriptors} instead of {@link com.google.protobuf.DescriptorProtos}.
  * The former descriptors provide a more powerful API.
  */
+//TODO:2018-10-29:dmytro.grankin: the class should be moved to `io.spine.code.proto`,
+// but the package already contains `FieldDeclaration`.
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/validate/FieldDeclaration.java]: Please resolve the issue within this PR.

File: base/src/main/java/io/spine/validate/FieldValue.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+ * <p>The exact type of the value is unknown since it is set
+ * by a user using a generated validating builder.
+ *
+ * <p>Map fields are considered in a special way and only values are validated.
+ * Keys don't require validation since they are of primitive types.
+ *
+ * @see <a href="https://developers.google.com/protocol-buffers/docs/proto3#maps">
+ *         Protobuf Maps</a>
+ */
+class FieldValue {
```
Comment (dmdashenkov) [inline base/src/main/java/io/spine/validate/FieldValue.java]: Can this be final?

File: base/src/test/java/io/spine/validate/FloatFieldValidatorTest.java [hunk 7]
```diff
@@ -20,14 +20,13 @@
...
 package io.spine.validate;
 
-import com.google.common.collect.ImmutableList;
 import org.junit.jupiter.api.DisplayName;
 
 @DisplayName("FloatFieldValidator should")
 class FloatFieldValidatorTest extends NumberFieldValidatorTest<Float, FloatFieldValidator> {
 
     private static final float HALF = 0.5F;
     FloatFieldValidatorTest() {
```
Comment (dmdashenkov) [inline base/src/test/java/io/spine/validate/FloatFieldValidatorTest.java]: Please add an empty line.

File: base/src/main/java/io/spine/code/proto/FieldDeclaration.java (hunk 1 above)
Comment (dmytro-grankin) [reply base/src/main/java/io/spine/code/proto/FieldDeclaration.java]: The index is zero-based. It is **not** about the number in the Proto declaration (`string field = 1`).

**Comment by dmytro-grankin** [comment]
@dmdashenkov PTAL.

**Pull request merged**